crates/loro-internal/src/kv_store/mem_store.rs

```diff
@@ -0,0 +1,395 @@
+//! │ MemKVStore                                                                                      │
+//! │┌ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ┐│
+//! │  Magic Number │ Schema Version │ Block Chunk   ...  │  Block Chunk    Block Meta │ Meta Offset  │
+//! ││     u32      │       u8       │    bytes    │      │     bytes     │   bytes    │     u32     ││
+//! │ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ │
+//! └─────────────────────────────────────────────────────────────────────────────────────────────────┘
+//! ```
+//!
+//! 1. Magic Number (4 bytes): A fixed value "LORO" to identify the file format.
+//! 2. Schema Version (1 byte): The version of the MemKVStore schema.
+//! 3. Block Chunks: A series of data blocks containing key-value pairs.
+//! 4. Block Meta: Metadata for all blocks, including block offset, first key, is_large flag, and last key if not large.
+//! 5. Meta Offset (4 bytes): The offset of the Block Meta section from the beginning of the file.
+//!
+//! ## Block Types
+//!
+//! There are two types of blocks: Normal Blocks and Large Value Blocks.
+//!
+//! ### Normal Block
+//!
+//! Normal blocks store multiple key-value pairs with compressed keys.
+//!
+//! ```
+//! ┌────────────────────────────────────────────────────────────────────────────────────────────┐
+//! │Block                                                                                   │
+//! │┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─┌ ─ ─ ─┌ ─ ─ ─ ┬ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ │
+//! │ Key Value Chunk  ...  │Key Value Chunk  offset │ ...  │ offset  kv len │Block Checksum││
+//! ││     bytes     │      │     bytes     │  u16   │      │  u16  │  u16   │     u32       │
+//! │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ┘│
+//! └────────────────────────────────────────────────────────────────────────────────────────────┘
+//! ```
+//!
+//! Each Key Value Chunk is encoded as follows:
+//!
+//! ```
+//! ┌───────────────────────────────────────────────────────────────┐
+//! │  Key Value Chunk                                              │
+//! │┌ ─ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ┬ ─ ─ ─ ┐│
+//! │ common prefix len key suffix len│key suffix│value len  value  │
+//! ││       u8        │     u16      │  bytes   │   u16   │ bytes ││
+//! │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ │
+//! └───────────────────────────────────────────────────────────────┘
+//! ```
+//!
+//! Encoding:
+//! 1. Compress key-value pairs data as Key Value Chunk.
+//! 2. Write offsets for each key-value pair.
+//! 3. Write the number of key-value pairs.
+//! 4. **Compress** the entire block using LZ4.
+//! 5. Calculate and append CRC32 checksum.
+//!
+//! Decoding:
+//! 1. Verify the CRC32 checksum.
+//! 2. **Decompress** the block using LZ4.
```

Comment:
Shuold we have a flag for the compression algorithm used?

Reply:
I think the results of different compression algorithms may not be very different, and we will probably not expose the param to the user, right?   

But we could use the first 7 bits of `is_large` for other purposes.